amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -114,12 +126,115 @@ public void restartFrom(RecordsRetrieved recordsRetrieved) {
         }
     }
 
+    @Override
+    public void notify(RecordsDeliveryAck recordsDeliveryAck) {
+        synchronized (lockObject) {
+            CompletableFuture<RecordFlow> triggeringFlowFuture = new CompletableFuture<>();
+            try {
+                evictAckedEventAndScheduleNextEvent(recordsDeliveryAck, triggeringFlowFuture);
+            } catch (Throwable t) {
+                errorOccurred(triggeringFlowFuture.getNow(null), t);
+            }
+            final RecordFlow triggeringFlow = triggeringFlowFuture.getNow(null);
+            if (triggeringFlow != null) {
+                updateAvailableQueueSpaceAndRequestUpstream(triggeringFlow);
+            }
+        }
+    }
+
+    // This method is not thread-safe. You need to acquire a lock in the caller in order to execute this.
+    @VisibleForTesting
+    void evictAckedEventAndScheduleNextEvent(RecordsDeliveryAck recordsDeliveryAck,
+            CompletableFuture<RecordFlow> triggeringFlowFuture) {
+        // Peek the head of the queue on receiving the ack.
+        // Note : This does not block wait to retrieve an element.
+        final RecordsRetrievedContext recordsRetrievedContext = recordsDeliveryQueue.peek();
+        // Check if the ack corresponds to the head of the delivery queue.
+        if (recordsRetrievedContext != null && recordsRetrievedContext.getRecordsRetrieved().batchUniqueIdentifier()
+                .equals(recordsDeliveryAck.batchUniqueIdentifier())) {
+            // It is now safe to remove the element
+            recordsDeliveryQueue.poll();
+            // Take action based on the time spent by the event in queue.
+            takeDelayedDeliveryActionIfRequired(shardId, recordsRetrievedContext.getEnqueueTimestamp(), log);
+            // Update current sequence number for the successfully delivered event.
+            currentSequenceNumber = ((FanoutRecordsRetrieved)recordsRetrievedContext.getRecordsRetrieved()).continuationSequenceNumber();
+            // Update the triggering flow for post scheduling upstream request.
+            triggeringFlowFuture.complete(recordsRetrievedContext.getRecordFlow());
+            // Try scheduling the next event in the queue, if available.
+            if (recordsDeliveryQueue.peek() != null) {
+                subscriber.onNext(recordsDeliveryQueue.peek().getRecordsRetrieved());
+            }
+        } else {
+            // Check if the mismatched event belongs to active flow. If publisher receives an ack for a
+            // missing event in active flow, then it means the event was already acked or cleared
+            // from the queue due to a potential bug.
+            if (flow != null && recordsDeliveryAck.batchUniqueIdentifier().getFlowIdentifier()
+                    .equals(flow.getSubscribeToShardId())) {
+                log.error(
+                        "{}: KCL BUG: Publisher found mismatched ack for subscription {}  ",
```

Comment:
Yha we can add more context. Chose to have "KCL Bug" so that the customers can reachout to us in this case. anyways we can convey this using "Unexpected exception" as well